ARCs/arc-draft_plugin_abstraction.md

```diff
@@ -0,0 +1,177 @@
+---
+arc: <to be assigned>
+title: Plugin-Based Account Abstraction
+description: An extendable standard for account abstraction using stateful applciations
+author: Joe Polny (@joe-p), Kyle Breeding aka krby.algo (@kylebeee)
+discussions-to: <URL>
+status: Draft
+type: Standards Track
+category: ARC
+created: 2024-01-08
+requires: 4
+---
+
+## Abstract
+This ARC proposes a standard for using stateful applications and rekey transactions to enable account abstraction on Algorand. The abstracted account is controlled by a single stateful application which is the auth address of the abstracted account. Other applications can be used as plugin to provide additional functionality to the abstracted account.
+
+## Motivation
+Manually signing transactions for every dApp interaction can be rather fatiguing for the end-user, which results in a frustrating UX. In some cases, it makes specfific app designs that require a lot of transactions borderline impossible without delegation or an escrow account.
+
+Another common point of friction for end-users in the Algorand ecosystem is ASA opt-in transactions. This is a paticularly high point of friction for onboarding new accounts since they must be funded and then initiate a transaction. This standard can be used to allow mass creation of non-custodial accounts and trigger opt-ins on their behalf.
+
+## Specification
+
+### Definitions
+**Abstracted Account** - An account that has functionality beyond a typical keypair-based account.
+
+**Abstracted Account App** - The stateuful application used to control the abstracted account. This app's address is the `auth-addr` of the abstracted account.
+
+**Plugin** - An additional application that adds functionality to the **Abstracted Account App** (and thus the **Abstracted Account**).
+
+**Admin** - An account, sepererate from the **Abstracted Account**, that controls the **Abstracted Account App**. In patiular, this app can initiate rekeys, add plugins, and transfer admin.
+
+
+### ARC4 Methods
+
+An abstracted app account that adheres to this standard **MUST** implement the following methods
+
+```json
+  "methods": [
+    {
+      "name": "createApplication",
+      "desc": "Create an abstracted account application",
+      "args": [
+        {
+          "name": "address",
+          "type": "address",
+          "desc": "The address of the abstracted account. If zeroAddress, then the address of the contract account will be used"
+        },
+        {
+          "name": "admin",
+          "type": "address",
+          "desc": "The admin for this app"
+        }
+      ],
+      "returns": {
+        "type": "void"
+      }
+    },
+    {
+      "name": "verifyAppAuthAddr",
+      "desc": "Verify the abstracted account is rekeyed to this app",
+      "args": [],
+      "returns": {
+        "type": "void"
+      }
+    },
+    {
+      "name": "rekeyTo",
+      "desc": "Rekey the abstracted account to another address. Primarily useful for rekeying to an EOA.",
```

Comment:
EOA?

Reply:
Externally Owned Account. Added definition in f797187